### PR TITLE
Fix `*.d.ts` generation race condition

### PR DESCRIPTION
**Pull request by coopbri**
##### Task link: https://trello.com/c/jJ3HW0My/156-dev-performance-optimization

- Fixed race condition caused by `tsup` `onSuccess` hook triggering before `*.d.ts` generation finishes
  - Improved previous solution of flaky sleep syscalls
- Simplified list of external packages for `tsup` build
- Added [dependency cycle](https://en.wikipedia.org/wiki/Circular_dependency) lint rule

## Test Steps

- Verify local build consistently works: CJS & ESM → DTS → local package publish

### CODE REVIEW

File: src/components/client/core/Accordion/Accordion.tsx [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import { FiChevronDown as ChevronDownIcon } from "react-icons/fi";
 
-import { Button, Icon } from "components/client";
```
Comment (coopbri) [inline src/components/client/core/Accordion/Accordion.tsx:3]: Cyclic dep

File: src/components/client/core/Drawer/Drawer.tsx [hunk 2]
```diff
@@ -1,7 +1,8 @@
 import { Portal } from "@ark-ui/react";
 import { FiX as CloseIcon } from "react-icons/fi";
 
-import { Button, Icon } from "components/client";
```
Comment (coopbri) [inline src/components/client/core/Drawer/Drawer.tsx:4]: Cyclic dep

File: src/components/client/core/Modal/Modal.tsx [hunk 3]
```diff
@@ -1,7 +1,8 @@
 import { Portal } from "@ark-ui/react";
 import { FiX as CloseIcon } from "react-icons/fi";
 
-import { Button, Icon } from "components/client";
```
Comment (coopbri) [inline src/components/client/core/Modal/Modal.tsx:4]: Cyclic dep

File: tsup.config.ts [hunk 4]
```diff
@@ -16,10 +30,11 @@ const tsupConfig = defineTsupConfig({
   sourcemap: true,
   minify: true,
   clean: true,
-  dts: true,
   splitting: false,
   format: ["cjs", "esm"],
-  external: ["react", "@ark-ui/react", "react-icons", "next"],
+  // NB: `peerDeps`, among others, are excluded (marked external) by default
+  // see https://tsup.egoist.dev/#excluding-packages
+  external: ["@ark-ui/react", "react-icons", "next"],
```
Comment (coopbri) [inline tsup.config.ts:37]: Verified bundle sizes were the exact same before and after removing `react`

Comment (hobbescodes) [reply tsup.config.ts:37]: Ahhh does this change have something to do with the addition of `next`? Just out of curiosity.

Comment (coopbri) [reply tsup.config.ts:37]: Nah, just an oversight on my part, `tsup` automatically excludes peer deps

Comment (hobbescodes) [reply tsup.config.ts:37]: Gotcha!

**Pull request merged**